velox/functions/prestosql/tests/DecimalArithmeticTest.cpp

```diff
@@ -105,72 +105,136 @@ TEST_F(DecimalArithmeticTest, add) {
   testDecimalExpr<TypeKind::SHORT_DECIMAL>(
-      expectedShortWithNulls,
+      makeNullableShortDecimalFlatVector(
+          {2, 4, std::nullopt, std::nullopt, std::nullopt}, DECIMAL(11, 3)),
       "minus(c0, c1)",
-      {shortWithNullsA, shortWithNullsB});
+      {makeNullableShortDecimalFlatVector(
+           {3, 6, std::nullopt, 6, std::nullopt}, DECIMAL(10, 3)),
+       makeNullableShortDecimalFlatVector(
+           {1, 2, 5, std::nullopt, std::nullopt}, DECIMAL(10, 3))});
+}
+
+TEST_F(DecimalArithmeticTest, decimalMultiplyTest) {
+  auto shortFlat = makeShortDecimalFlatVector({1000, 2000}, DECIMAL(17, 3));
+  // Multiply short and short, returning long.
+  testDecimalExpr<TypeKind::LONG_DECIMAL>(
+      makeLongDecimalFlatVector({1000000, 4000000}, DECIMAL(34, 6)),
+      "multiply(c0, c1)",
+      {shortFlat, shortFlat});
+  // Multiply short and long, returning long.
+  auto longFlat = makeLongDecimalFlatVector({1000, 2000}, DECIMAL(20, 3));
+  auto expectedLongFlat =
+      makeLongDecimalFlatVector({1000000, 4000000}, DECIMAL(37, 6));
+  testDecimalExpr<TypeKind::LONG_DECIMAL>(
+      expectedLongFlat, "multiply(c0, c1)", {shortFlat, longFlat});
+  // Multiply long and short, returning long.
+  testDecimalExpr<TypeKind::LONG_DECIMAL>(
+      expectedLongFlat, "multiply(c0, c1)", {longFlat, shortFlat});
+
+  // Multiply long and long, returning long.
+  testDecimalExpr<TypeKind::LONG_DECIMAL>(
+      makeLongDecimalFlatVector({1000000, 4000000}, DECIMAL(38, 6)),
+      "multiply(c0, c1)",
+      {longFlat, longFlat});
+
+  // Multiply short and short, returning short.
+  shortFlat = makeShortDecimalFlatVector({1000, 2000}, DECIMAL(6, 3));
+  testDecimalExpr<TypeKind::SHORT_DECIMAL>(
+      makeShortDecimalFlatVector({1000000, 4000000}, DECIMAL(12, 6)),
+      "c0 * c1",
+      {shortFlat, shortFlat});
+  auto expectedConstantFlat =
+      makeShortDecimalFlatVector({100000, 200000}, DECIMAL(9, 5));
+
+  // Constant and Flat arguments.
+  testDecimalExpr<TypeKind::SHORT_DECIMAL>(
+      expectedConstantFlat, "1.00 * c0", {shortFlat});
+
+  // Flat and Constant arguments.
+  testDecimalExpr<TypeKind::SHORT_DECIMAL>(
+      expectedConstantFlat, "c0 * 1.00", {shortFlat});
+
+  // Long decimal limits
+  try {
+    testDecimalExpr<TypeKind::LONG_DECIMAL>(
+        {},
+        "c0 * 10.00",
+        {makeLongDecimalFlatVector(
+            {std::numeric_limits<int128_t>::max()}, DECIMAL(38, 0))});
+    FAIL();
+  } catch (VeloxUserError& ex) {
+    EXPECT_EQ(
+        ex.message(),
+        "integer overflow: 170141183460469231731687303715884105727 * 1000");
+  }
+  try {
+    testDecimalExpr<TypeKind::LONG_DECIMAL>(
+        {},
+        "c0 * 10.00",
+        {makeLongDecimalFlatVector(
+            {std::numeric_limits<int128_t>::min()}, DECIMAL(38, 0))});
+    FAIL();
+  } catch (VeloxUserError& ex) {
+    EXPECT_EQ(
+        ex.message(),
+        "integer overflow: -170141183460469231731687303715884105728 * 1000");
```

Comment:
This error will change since we specialize CheckedArithmetic for decimal here that is now merged. https://github.com/facebookincubator/velox/pull/2163/files#diff-ba972f0854a87c036d8b3f4e749e3936b0e74fbd0fa4f3c92a647e99d94f2c5b

Reply:
Here we are multiplying the int128_t numbers not the wrapper structs.